lib/pbench/test/unit/server/database/test_server_config_db.py

```diff
@@ -1,56 +1,309 @@
+        """
+        q = FakeQuery(self)
+        self.queries.append(q)
+        return q
+
+    def add(self, config: ServerConfig):
+        """
+        Add a DB object to a list for testing
+
+        Args:
+            config: A server configuration setting object
+        """
+        self.added.append(config)
+
+    def commit(self):
+        """
+        Mock a commit operation on the DB session. If the 'raise_on_commit'
+        property has been set, "fail" by raising the exception. Otherwise,
+        mock a commit by updating any cached "committed" values if the "known"
+        proxy objects have changed, and record any new "added" objects.
+        """
+        if self.raise_on_commit:
+            raise self.raise_on_commit
+        for k in self.known:
+            if self.committed[k].value != self.known[k].value:
+                self.committed[k].value = self.known[k].value
+        for a in self.added:
+            a.id = self.id
+            self.id += 1
+            self.known[a.id] = a
+            self.committed[a.id] = FakeRow(id=a.id, key=a.key, value=a.value)
+        self.added = []
+        self.deleted = []
+
+    def rollback(self):
+        """
+        Just record that rollback was called, since we always raise an error
+        before changing anything during the mocked commit.
+        """
+        self.rolledback += 1
+
+
 class TestServerConfig:
-    def test_bad_key(self, db_session):
+    session = None
+
+    def check_session(self, added=[], queries=0, committed=[], rolledback=0):
+        """
+        Encapsulate the common checks we want to make after running test cases.
+
+        Args:
+            added: A list of object proxies we expect to have been added (but
+                not yet committed)
+            queries: A count of queries we expect to have been made
+            committed: A set of objects we expect to have been committed
+            rolledback: True if we expect rollback to have been called
+        """
+        session = self.session
+        assert session
+        assert len(session.queries) == queries
+        assert len(session.added) == len(added)
+        xlated = [
+            a
+            for a in session.added
+            for k in added
+            if a.key == k.key and a.value == k.value
+        ]
+        assert session.added == xlated
+        assert sorted(list(session.committed.values())) == sorted(committed)
+        assert session.rolledback == rolledback
+
+    @pytest.fixture(autouse=True, scope="function")
+    def fake_db(self, monkeypatch, server_config):
+        """
+        Fixture to mock a DB session for testing.
+
+        We patch the SQLAlchemy db_session to our fake session. We also store a
+        server configuration object on the Database.Base. (TODO: I was unable
+        to monkeypatch this due to the odd context of our base object
+        modifications during DB initialization; this is "unsatisfying" but
+        shouldn't be a problem)
```

Comment:
I don't understand the `TODO` comment:  leaving aside the fact that it doesn't seem to indicate or describe anything which we need to do, it looks to me like the code is successfully using `monkeypatch`, and it's not obvious to me what is unsatisfying about your solution.

Consider changing "TODO" to "NOTE", or, better yet, just removing the whole parenthetical?

(FWIW, I _think_ I **like** the `scope="function"` and `autouse=True`, although it elevates things to a new level of "magic", since the setup is now almost completely invisible.)

Reply:
Pytest is all about the magic.

The TODO refers to the `Database.Base.config = server_config`, which was described in the previous sentence. Loading attributes onto SQLAlchemy's `declarative_base` was always a bit of a hack, but pytest doesn't like trying to monkeypatch this attribute since it doesn't "really" exist.

So, sure, maybe `NOTE` would be a bit better, but I was hoping there *was* something "to do" here and I just hadn't figured it out yet...

Reply:
> The TODO refers to the `Database.Base.config = server_config`

Ah!  I do so hate pronouns.  Sometimes I think they cause more trouble than they alleviate (said using three 😉).

As for the other, I'm inclined to say that, if it seems to work, it's probably good enough.  🍤